### PR TITLE
Made necessary changes for porting over to pytorch lightning

### PR DESCRIPTION
**Pull request by anshulcgm**
- Ported over all modules from nn.Module to pl.LightningModule
- Wrote Data Module for configuring dataloaders
- Setup.py file to install spectr as a package
- spectr_config.py so we can use yaml files and keep track of different experiments
- Wrote training step, validation step and configure optimizers method in custom_segmenter.py file 
- Need labels to be in [batch_size, height, width] format with each pixel representing the class it belongs too
- Not fully operational yet but should be after output data is in the correct format

### CODE REVIEW

File: spectr/spectr_config.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
+from typing import Dict, Any
+
+import torch
+import torch.nn as nn
+from spectr.src.custom_segmenter import CustomSegmenter
+from spectr.src.cityscapes_data_module import CityScapesDataModule
+
+class SpectrConfig:
+    def __init__(self, train_dataset_path: str, train_labels_path: str, val_dataset_path: str, val_labels_path: str, test_dataset_path: str, test_labels_path: str, encoder_config: Dict[str, Any], decoder_config: Dict[str, Any], batch_size: int, learning_rate: float, training_steps: int, gamma: float, accelerator: str):
```
Comment (adisingh50) [inline spectr/spectr_config.py:9]: add a docstring comment here

File: spectr/src/cityscapes_data_module.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        """Initializes DataModule for Cityscapes dataset that returns associated dataloaders
+
+        Args:
+            data_dir: 
+        """
+        super().__init__()
+
+        self.train_data_dir = train_data_dir
+        self.train_labels_dir = train_labels_dir
+        self.vaL_data_dir = vaL_data_dir
```
Comment (adisingh50) [inline spectr/src/cityscapes_data_module.py]: small typo here. change `L` to `l`

File: spectr/src/custom_segmenter.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
+from typing import Dict, List, Any
+import pdb
+
+import torch
+import torch.nn as nn
+import pytorch_lightning as pl
+from encoder_network import EncoderNetwork
+from masked_decoder import MaskedTransformer
```
Comment (adisingh50) [inline spectr/src/custom_segmenter.py]: now that we are importing spectr as a module, let's change lines 7 and 8 to:

`from spectr.src.encoder_network import EncoderNetwork`
`from spectr.src.masked_decoder import MaskedTransformer`

File: spectr/src/custom_segmenter.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    def training_step(self, curr_batch: List[torch.Tensor], batch_idx: int) -> float:
+        """Training step method called by Pytorch Lightning
+        
+        Args:
+            curr_batch: A list of tensors (x, y). x is a tensor of shape [batch_size, num_channels, height, width], y is a tensor of shape [batch_size, height, width]
+        
+        Returns:
+            output_loss: A number indicating cross entropy loss 
+        """
+        pdb.set_trace()
```
Comment (adisingh50) [inline spectr/src/custom_segmenter.py]: probably remove this so it doesn't interfere with actual training?

File: spectr/src/custom_segmenter.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+            training_steps: Number of batches for model to process during training
+            gamma: Decay rate
+        """
+        super(CustomSegmenter, self).__init__()
+
+        self.encoder = EncoderNetwork(**encoder_config)
+        self.decoder = MaskedTransformer(**decoder_config)
+        self.learning_rate = learning_rate
+        self.training_steps = training_steps
+        self.gamma = gamma
```
Comment (adisingh50) [inline spectr/src/custom_segmenter.py:29]: use this in StepLR

**Review by adisingh50 — CHANGES_REQUESTED** [review]
just some small comments. pytorch lightning seems very high iq

**Review by adisingh50 — APPROVED** [review]
LGTM
